x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authc/jwt/JwtAuthenticationToken.java

```diff
@@ -0,0 +1,108 @@
+ * 2.0; you may not use this file except in compliance with the Elastic License
+ * 2.0.
+ */
+package org.elasticsearch.xpack.security.authc.jwt;
+
+import com.nimbusds.jose.JWSHeader;
+import com.nimbusds.jose.util.Base64URL;
+import com.nimbusds.jwt.JWTClaimsSet;
+import com.nimbusds.jwt.SignedJWT;
+
+import org.elasticsearch.common.settings.SecureString;
+import org.elasticsearch.core.Nullable;
+import org.elasticsearch.xpack.core.security.authc.AuthenticationToken;
+import org.elasticsearch.xpack.security.authc.BearerToken;
+
+import java.text.ParseException;
+import java.util.concurrent.atomic.AtomicReference;
+
+/**
+ * An {@link AuthenticationToken} to hold JWT authentication related content.
+ */
+public class JwtAuthenticationToken extends BearerToken {
+    // Stored members (Note: also includes super.bearerString, which is a SecureString too)
+    protected final SecureString clientAuthorizationSharedSecret; // optional, nullable
+
+    // Processed members (ParseException)
+    protected final AtomicReference<SignedJWT> signedJwt = new AtomicReference<>(null);
+    protected final AtomicReference<JWSHeader> jwsHeader = new AtomicReference<>(null);
+    protected final AtomicReference<JWTClaimsSet> jwtClaimsSet = new AtomicReference<>(null);
+    protected final AtomicReference<Base64URL> jwtSignature = new AtomicReference<>(null);
+
+    /**
+     * Store a mandatory JWT and optional Shared Secret. Parse the JWT, and extract the header, claims set, and signature.
+     * Throws IllegalArgumentException if bearerString is missing, or if JWT parsing fails.
+     * @param bearerString Base64Url-encoded JWT for End-user authorization. Required by all JWT realms.
+     * @param sharedSecret Base64Url-encoded Shared Secret for Client authorization. Required by some JWT realms.
+     */
+    public JwtAuthenticationToken(final SecureString bearerString, @Nullable final SecureString sharedSecret) {
+        super(bearerString); // super.bearerString
+        if (bearerString == null) {
+            throw new IllegalArgumentException("JWT bearer token must be non-null");
+        }
+        this.clientAuthorizationSharedSecret = sharedSecret; // optional, nullable
+        // Parse JWT
+        try {
+            final SignedJWT parsed = SignedJWT.parse(bearerString.toString());
+            this.signedJwt.set(parsed);
+            this.jwsHeader.set(parsed.getHeader());
+            this.jwtClaimsSet.set(parsed.getJWTClaimsSet());
+            this.jwtSignature.set(parsed.getSignature());
+        } catch (ParseException e) {
+            this.signedJwt.set(null);
+            this.jwsHeader.set(null);
+            this.jwtClaimsSet.set(null);
+            this.jwtSignature.set(null);
+            throw new IllegalArgumentException("Failed to parse JWT bearer token", e);
+        }
+    }
+
+    @Override
+    public String principal() {
+        return super.bearerString.toString();
```

Comment:
Principal claim is per JWT realm, so JwtAuthenticationToken has no way to predict what claim to use for the principal. That is true if there is one JWT realm or many.

For now, I will return a constant like most other token-based realms (ex: BearerToken, OpenIdConnectToken, SamlToken, KerberosToken). Caching comes up later in the meta issue, so we can revisit at that point.

Reply:
When we get to that discussion, some of the alternatives I was considering were:
- null
- constant string
- Digest of the JWT (TBD: SHA256 vs PBKDF2 vs other)
- Truncated digest of the JWT (ex: Git allows shortening SHA1 hashes from 40 hex characters down to 8)

Hopefully these could address leaking the JWT in audit logs and authentication failure messages. For caching of frequently changing JWTs, that might be better to address inside each JWT realm instance, because each one knows the principal claim they need to use.

Reply:
TODO Assumption is issuer and audiences are mandatory, but subject is optional. OIDC ID Tokens say subject is required, but we allow OIDC/JWT realms to pick a different claim for the ES principal, so theoretically we could potentially treat subject as optional.

For cache key, we can do something like this:
1. If subject is required and/or present, then issuer+audiences+subject.
2. If subject is optional and missing, then we can strip the header and footer, strip out nonce and date fields, sort the renaming claims, and use that as an opaque key. We agreed logging that in audit logs would be fine because logging only part of the claimset and no header/footer is safe enough.

Reply:
Done. I applied both approaches. JwtAuthenticationToken.principal() now returns one of these two:
- If sub claim is present, principal is iss/aud/sub.
- If sub claim is absent, principal is iss/aud/orderedClaims; claims stripped from orderedClaims: iss, aud, exp, iat, nbf, auth_time, nonce, jti

We discussed and it is OK to log the remaining claims. Followups I am wondering about:
- Are there any other claims I should remove from orderedClaims
- We didn't discuss what to return for credentials(), I am just returning the same as principal() wrapped in a SecureString, so let me know if you think that is OK.

Reply:
@tvernum let me know what you think of the changes, and the two follow up questions.

Reply:
I think it's overkill. We say that the JWT needs to conform to OIDC ID Tokens, so `sub` is required. If it's missing we could just set it to `"(no-subject)"` and get a good enough result.
I don't think we _need_ to rip out what's there, but it's substantially more complexity for very little value.

Reply:
> We didn't discuss what to return for credentials(), I am just returning the same as principal() wrapped in a SecureString, so let me know if you think that is OK.

I suspect we should just drop the `credentials` method from `AuthenticationToken`, but that's a separate conversation/PR.
I would lean towards `endUserSignedJwt` being the credential, but it's a bit arbitrary either way.

Reply:
Resolved. Switched to endUserSignedJwt for credentials().